src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.RenameDocumentUndoUnit.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.OLE.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    internal partial class VisualStudioWorkspaceImpl
+    {
+        private class RenameDocumentUndoUnit : IOleUndoUnit
+        {
+            private readonly VisualStudioWorkspaceImpl _workspace;
+            private readonly DocumentId _documentId;
+            private readonly string _fromName;
+            private readonly string _toName;
+
+            public RenameDocumentUndoUnit(VisualStudioWorkspaceImpl workspace, DocumentId documentId, string fromName, string toName)
+            {
+                _workspace = workspace;
+                _documentId = documentId;
+                _fromName = fromName;
+                _toName = toName;
+            }
+
+            public void Do(IOleUndoManager pUndoManager)
+            {
+                var updatedSolution = _workspace.CurrentSolution.WithDocumentName(_documentId, _toName);
+                _workspace.TryApplyChanges(updatedSolution);
```

Comment:
i can't remember how undo units work, but don't you need to add back the "redo" action back into the stack?  @jasonmalinowski might know?  You can look at teh work i did with nuget-package-add to see how this can be done.